### DISCUSSION TITLE
Missing Languages: Swiss language codes (fr, de ,it)

### DISCUSSION THREAD

**Discussion by josias-r** [directus/directus#16665] [Feature Requests]
- de-CH -> German (Switzerland)
- fr-CH -> French (Switzerland)
- it-CH -> Italian (Switzerland)

The above languages are all missing from the languages dropdown.

**Comment by azrikahar** [comment]
With this concern: https://github.com/directus/directus/pull/14858#issuecomment-1206791665 in mind, would you be open to provide the translations when they are set up? 

Labeling this with `external` for now as I believe it will need to be enabled in Crowdin first before we can add it to https://github.com/directus/directus/blob/main/app/src/lang/available-languages.yaml.

**Comment by josias-r** [comment]
@azrikahar Well 99% of the time the translations are the same as the corresponding langauges (french, italian, german). So translating it would be fairly easy, I can do that.

It's only really differences in typography such as different qote characters, etc

But: I created this ticket because I was trying to add a [translation string](https://docs.directus.io/configuration/translation-strings.html) myself, without thinking about the translations you provide out of the box for directus.

So, for what reason are these translations the same? Shouldn't it be one thing to have directus translated in the handfull of languages you think makes sense, but still give the users the possiblity to add translations in any language they want?

**Comment by azrikahar** [comment]
> I created this ticket because I was trying to add a [translation string](https://docs.directus.io/configuration/translation-strings.html) myself

Oh that's a pretty important context here! Thanks for clarifying 👍

> So, for what reason are these translations the same?  Shouldn't it be one thing to have directus translated in the handfull of languages you think makes sense, 

Unfortunately I can only say I have no idea either. I can only assume it's partly to make them usable by vue-i18n, and possibly the notion of "any language" wasn't included in the MVP when locale support was first added 🙂

> but still give the users the possiblity to add translations in any language they want?

Feels like this was brought up before in a past discussion but I can't seem to find it, would be a nice improvement indeed.

That being said, if the goal is to add a translation string for a language that does not exist in the language dropdown, I believe you can edit the raw value to your desired language code in the mean time:

https://user-images.githubusercontent.com/42867097/204249657-55d68e5e-c570-445e-9cfb-aa52937087ab.mp4

Out of curiosity, would you mind elaborating further on this use case? are you using the translations elsewhere outside of the app?

**Comment by josias-r** [comment]
> That being said, if the goal is to add a translation string for a language that does not exist in the language dropdown, I believe you can edit the raw value to your desired language code in the mean time:

Ah cool, thank you for showing a workaround!

> Out of curiosity, would you mind elaborating further on this use case? are you using the translations elsewhere outside of the app?

No I was not thinking of using the translations outside the app. The use-case would really only be to have the Collections translated in the UI of directus.

To be perfectly honest, I stumbeled upon this with a different goal in mind - to have translatable content in a collection in directus. I came across this in your docs showing of that possiblitly:
<img width="716" alt="Screenshot 2022-11-28 at 11 16 52" src="https://user-images.githubusercontent.com/11424820/204253324-3fe8d507-2f7b-4fc2-9a83-bc2113659fb7.png">
But as it turns out this showcase was actually just about translating that note marked in red in the Screenshot. Not about the big purple "Translations" of content in the middle of the screen, which is a bit misleading.

So I have two suggestions:
- Create a new documentation page about the translation component (a collection setup to support add
ing data in multiple languages). Right now there seems to be no such documentation: 
<img width="576" alt="Screenshot 2022-11-28 at 11 23 39" src="https://user-images.githubusercontent.com/11424820/204254368-f1ac673e-244d-453e-81bf-dee5fcd37b3a.png">

- Clearly differentiate in your documentation on which one is which:
  - App translations / Crowdin -> translations used by the Directus FrontEnd
  - Translation Strings -> custom translations available to be used in the Directus FrontEnd for Collection and Fields created by an admin
  - New*: Translation Component for collections -> Used for having a data of a collection in multiple languages (example: blog post translatable into multiple languages)

**Comment by paescuj** [comment]
>> So, for what reason are these translations the same? Shouldn't it be one thing to have directus translated in the handfull of languages you think makes sense,

> Unfortunately I can only say I have no idea either. I can only assume it's partly to make them usable by vue-i18n, and possibly the notion of "any language" wasn't included in the MVP when locale support was first added 🙂

Yes, I think it has been implemented like this for the sake of simplicity rather than any specific reasons.
The data is currently taken from [`available-languages.yaml`](https://github.com/directus/directus/blob/62b4a2413ca310b1bad23c64fe37e8feac710c1e/app/src/lang/available-languages.yaml (which contains the most popular languages)
) and I do agree that it would make sense to use a different source for "in-app" translations, also to populate the "Languages Collection" table.

(Of course there is an NPM package for this -> https://www.npmjs.com/package/all-iso-language-codes 😉)

**Comment by azrikahar** [comment]
> Create a new documentation page about the translation component (a collection setup to support add
ing data in multiple languages). Right now there seems to be no such documentation:

> New*: Translation Component for collections -> Used for having a data of a collection in multiple languages (example: blog post translatable into multiple languages)

@josias-r I think https://docs.directus.io/configuration/data-model/relationships.html#translations-o2m was meant to address this as a similar question was brought up in the past, but I can also confirm searching "translations" doesn't seem to suggest it in Algolia search.

In regards to the suggestion of arranging it as a new page and/or new section, I do acknowledge that it can be invaluable, but I'll defer to @erondpowell as to what would be the best course of action to tackle this pain point 😉

> I do agree that it would make sense to use a different source for "in-app" translations, also to populate the "Languages Collection" table.

@paescuj I concur! Though I also wonder can we use a similar approach as described by that package to generate the json file, and since it seems like (at least based on the `available-languages.yaml` as mentioned) we're only looking for IETF BCP 47 language codes. 🤔

---

Moving this to discussion in the meantime while we can flesh out the action items (in terms of the codebase) a bit more first. We can track the docs additions/changes separate if needed.

**Reply by paescuj** [reply]
> @paescuj I concur! Though I also wonder can we use a similar approach as described by that package to generate the json file, and since it seems like (at least based on the available-languages.yaml as mentioned) we're only looking for IETF BCP 47 language codes. 🤔

Small correction: The data from `available-languages.yaml` is only used for "Translation Strings" while the data for the "Translation Collection" is currently hard-coded here: https://github.com/directus/directus/blob/d6b8e5963d955769cfe2a721014ca17191d8850c/app/src/modules/settings/routes/data-model/field-detail/store/alterations/translations.ts#L234-L275

If feasible, I would support generating the data ourselves. Note that for the "Translation Collection" we would also need the text direction besides the ISO 639-1 / IETF BCP 47 data.

**Reply by azrikahar** [reply]
> the data for the "Translation Collection" is currently hard-coded here

Ohhh that makes so much more sense. I was fixated on the supported locales in the yaml file and totally glanced over and missed the mention of the auto-generated Languages table 🙈 thanks for clarifying!

> If feasible, I would support generating the data ourselves. Note that for the "Translation Collection" we would also need the text direction besides the ISO 639-1 / IETF BCP 47 data.

Sounds great, especially since we can have finer control on the structure. Really nice point in regards to needing text direction as well 💯

**Reply by josias-r** [reply]
> @josias-r I think https://docs.directus.io/configuration/data-model/relationships.html#translations-o2m was meant to address this as a similar question was brought up in the past, but I can also confirm searching "translations" doesn't seem to suggest it in Algolia search.

Ah yes, that is exactly what I was expecting to find, but did not. First time seeing it now, very strange that Algolia doesn't pick up on it.

**Comment by erondpowell** [comment]
> App translations / Crowdin -> translations used by the Directus FrontEnd
Translation Strings -> custom translations available to be used in the Directus FrontEnd for Collection and Fields created by an admin
New*: Translation Component for collections -> Used for having a data of a collection in multiple languages (example: blog post translatable into multiple languages)

Yes, this is a known issue in queue in Docs Land. Lots of docs to write! 🏃 

Crowdin and translation strings docs are fairly-well flushed out. 

Until the docs are ready, you can find an implementation of the [Translations O2M Interface](https://docs.directus.io/configuration/data-model/relationships.html#translations-o2m) in the Community Cloud Demo project. You can go to `settings > data model > articles` and see how it is configured.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
